### ISSUE TITLE
Automated testing strategy in NewPipeExtractor

### ISSUE THREAD

**Issue by XiangRongLin** [TeamNewPipe/NewPipeExtractor#481]
Currently all CI builds on dev branch and PR are failing due to test failures. Then figuring out whether those are the usual suspects or caused by my own changes is a hassle and makes CI worthless.

Reason for that is the nature of unit tests relying on external services. In this case it seems Peertube is denying the connection. see https://travis-ci.org/github/TeamNewPipe/NewPipeExtractor/builds/749264136#L450
Since these tests are activly harmful, i would suggest ignoring them all for now.
There are also seem to be some SoundCloud test which are simply broken.

The long term solution is making the unit tests not rely on the external service. The expected websites would need to be saved locally as html and injected into the extractor in the tests. Additional benefits are that tests won't fail because the video/playlist/channel got deleted/changed.

A new problem that is introduced with this are missing tests for API changes of external services. For that a seperate test suite can be added, which only covers this (smoke test). Those should then NOT be run on every PR/Push but maybe nightly/weekly. Those test then also can have big buffers between executions to avoid getting blocked. Those jobs could also update the local html files.
Running without those short term is also fine since most API changes are breaking the extractor. Meaning the NewPipe repo will get bug reports depending on the feature (YT decrypt function, YT comments)

With this the unit tests are covering regressions and the smoke tests are covering whether the feature is actually working.

Thoughts?

---

In short:

1. - [x] ~~Ignore~~ Fix all peertube tests #511 #512
2. - [x] ~~Adjust architecture to allow injecting html into extractor~~ Missed that its already in place
3. - [x] Save websites needed for tests as html locally #482
   Functionality is implemented. See `YoutubeMixPlaylistExtractorTest` for usage
4. - [x] Adjust tests and inject html files into tests #513 #514 
5. - [x] ~~Implement smoke tests~~ Existing tests can be used with my approach. 
6. - [x] Add seperate CI job, which runs smoke tests nightly #516
7. - [x] Add option to mark tests as flaky. So they are not be run against real website in CI process.
  see comment https://github.com/TeamNewPipe/NewPipeExtractor/pull/516#issuecomment-761177779.
8. - [ ] Mark flaky tests

**Comment by Stypox** [comment]
I already did some efforts a while ago at mocking extractor tests on the har-mock branch on my fork, maybe it can be useful: https://github.com/Stypox/NewPipeExtractor/tree/har-mock

**Comment by XiangRongLin** [comment]
@Stypox 
Definetly interesting to see. Especially that a thing such as har exists.
For most har would be overkill i think, but there are some things where headers&co are important where that would be needed.

Do you have plans to get that merged or should i just strip out stuff that i want?

**Comment by Stypox** [comment]
That branch was just used to process HARs provided by users, so it is not meant as a real testing strategy. So feel free to take snippets of code from it ;-)
